### ISSUE TITLE
Storage library needs to use Shade plugin to prevent class path versioning issues

### ISSUE THREAD

**Issue by nicolehaugen79** [Azure/azure-storage-java#233]
We have noticed an issue with some of the Azure Mgmt SDKs in that they don't repackage their dependencies into a custom Microsoft namespace, using either the Maven Shade plugin or the Gradle Shadow plugin. Without using these plugins, this can cause versioning issues when this SDK is used in conjunction with other open source libraries using similar, but different versions of dependencies (e.g. 'class path hell'). Often, when class path hell occurs, you'll see an exception at run time saying "NoSuchMethod" because there are two versions of the same dependency used, but the wrong version gets loaded for one of the libraries that uses it. Using the Shade\Shadow plugins will prevent this.

I'm logging this issue because in looking at the POM for the azure-storage-java library, it doesn't reference the Maven Shade plugin. Please take a look at using this plugin to ensure that version conflicts don't occur when using the azure-storage-java SDK.

Here's a link to a similar issue that was logged against the Document DB library that describes this type of problem in greater detail since a versioning conflict issue was found when using the Document DB library in combination with the App Insights SDK. https://github.com/Azure/azure-documentdb-java/issues/90

**Comment by jofriedm-msft** [comment]
Hi @nicolehaugen79, this plugin seems like something that should be opted into by a user when creating an executable. As this is a library we do not want to unnecessarily repackage dependencies for all customers. There is also the legal concerns of repacking dependencies under the Microsoft namespace.

**Comment by nicolehaugen79** [comment]
I will follow up on these concerns that you have and will get back to you next week.

**Comment by dhaval24** [comment]
@jofriedm-msft this plugin can be used by customer if he has that dependency in his project. If the customer isn't aware of the dependency in the project then there is no way that he has idea to repackage this. In the event of this situation, when there are 2 external libraries which use different versions of same jars then this will cause class path conflict. It feels like the way to deal with this is to repackage very common jars. As far as legal concerns are there I think any OSS project which has a license which allows using it in any format and repackaging should be fine but I could be mistaken here. Also there might be other industry wide best practices which I might not be aware of for avoiding such conflicts. Feel free to share it if you know.

**Comment by mirobers** [comment]
@nicolehaugen79 @dhaval24 Thanks for the suggestions. Looking over this issue, I feel there's a tradeoff between potential versioning issues one side and package size and dependency management on the other. Using Shade unconditionally would appear to remove certain control from the user, namely the ability to swap in a different version of a dependent package.

I wonder if Shade could be applied by users only as needed for versioning and packaging, so the mvn hierarchy remains simple and the user has detailed control over their own product. This is something similar to the recommendation in one [stackoverflow answer](https://stackoverflow.com/a/13620420/3547195):

> Avoid using uber-jar as Maven dependency, as it is ruining the dependency resolution feature of Maven. Normally we create uber-jar only for the final artifact for actual deployment or for manual distribution, but not for putting to Maven repository.

If it is simple and straightforward to distribute both ways, I'd be interested in exploring that. But only distributing with Shade feels to me like we're losing more than we're gaining. Thoughts?

**Comment by dhaval24** [comment]
@mirobers I agree with your point. This is a classic trade of similar to what is "space time trade of in early computer science". I believe that shading should be done for only some of the vulnerable libraries which can possibly have very high probability of conversioning conflicts. Also, I have seen a common trend in the industry to shade the libraries which are famous for introducing breaking changes (popular example is Guava). For most of the other libraries, we should have a ubiquitous approach to use the version which has the highest downloads/users according to the maven central. Thoughts?

**Comment by yeroc** [comment]
+1

I'd suggest keeping the current artifacts as-is (unshaded) but publishing secondary shaded artifacts.  The simplest way is to configure the maven-shade-plugin to attach the shaded artifact under a different classifier (https://maven.apache.org/plugins/maven-shade-plugin/examples/attached-artifact.html) so that end-users can select whichever artifact is more appropriate to their situation.

**Comment by dhaval24** [comment]
@yeroc I am not sure what do you mean here. I think most common apache dependencies should always be shaded. Most popular projects do shade it. For example Guava is also almost always shaded. It  causes a lot of troubles, and our unshaded jars might replace what customers have.

**Comment by yeroc** [comment]
@dhaval24 I agree that in general people will want to use the shaded artifacts but if there are concerns about backwards compatibility then leave the main artifact as-is and attach a shaded artifact under a classifier and document that for people to use as an alternative.

**Comment by dhaval24** [comment]
@yeroc I think this is very contrary to what Java and engineering standard is in general. It would be very wiered to release two versions of same library with exactly same features:
1. With shaded jars
2. without shaded jars

IMO it leads to serious maintenance problems like -
1. Complicated build scripts to generate two different artifacts from same code base.
2. Very weird versioning standards (probably not adhering to semantic versioning).
3. Likelyhood of making simultaneous releases when the problem occurs in one version but not other too.

Most importantly this confuses our users more than ever (While they are midst transition). 

Regarding your comment on backward compatibilty, I do not see any potential issues why a shaded uber jar will create backward compatibility issues in any situations (if you can convince me with an example would be great)
